static/js/components/Loading.js

```diff
@@ -1,27 +1,48 @@
 // @flow
 import React from "react"
+import R from "ramda"
 
-import { anyProcessing, anyError, allLoaded } from "../util/rest"
-
-import type { RestState } from "../flow/restTypes"
-
-export default class Loading extends React.Component {
-  props: {
-    restStates: Array<RestState<*>>,
-    renderContents: Function
-  }
+type LoadingProps = {
+  loaded: boolean,
+  errored: boolean
+}
 
-  render() {
-    const { restStates, renderContents } = this.props
+const withLoading = R.curry((
+  shouldLoadData: (Object, Object) => boolean,
+  loadData: Object => void,
+  LoadedComponent: Class<React.Component<*, *, *>>
+) => {
+  return class extends React.Component {
+    props: LoadingProps
 
-    if (anyProcessing(restStates) || !allLoaded(restStates)) {
-      return <div>Loading</div>
+    componentWillMount() {
+      loadData(this.props)
     }
 
-    if (anyError(restStates)) {
-      return <div>Error loading page</div>
+    componentDidUpdate(prevProps: Object) {
+      if (shouldLoadData(prevProps, this.props)) {
+        loadData(this.props)
+      }
     }
 
-    return renderContents()
+    render() {
+      const { loaded, errored, ...props } = this.props
+
+      if (errored) {
+        return <div className="errored">Error loading page</div>
+      }
+
+      if (!loaded) {
+        return <div className="loading">Loading</div>
+      }
+
+      return (
+        <div className="loaded">
+          <LoadedComponent {...props} />
+        </div>
+      )
+    }
   }
-}
+})
+
+export default withLoading
```

Comment:
Neither `LoadedComponent.componentWillMount` nor `LoadedComponent.componentDidUpdate` will get called until the component mounts when `errored === false && loaded === true`, hence why I moved those implementations outside of the container. If we don't like this, I don't see the HOC approach being viable unless we hide the `LoadedComponent` with CSS or something hacky like that.